creusot-contracts/src/ghost_ptr.rs

```diff
@@ -0,0 +1,165 @@
+    #[ensures(result == (ptr1 == ptr2))]
+    pub fn are_eq(&self, ptr1: GhostPtr<T>, ptr2: GhostPtr<T>) -> bool
+    where
+        T: Sized,
+    {
+        self.injective_lemma();
+        ptr1.addr() == ptr2.addr()
+    }
+
+    /// Leaks memory iff the precondition fails
+    #[requires(self@.is_empty())]
+    pub fn drop(self) {}
+}
+
+impl<T: ?Sized> GhostPtrExt<T> for GhostPtr<T> {
+    #[trusted]
+    #[logic]
+    #[ensures(forall<t: GhostPtrToken<T>> !t@.contains(result))]
+    #[ensures(forall<ptr: GhostPtr<T>> ptr.addr_logic() == result.addr_logic() ==> ptr == result)]
+    fn null_logic() -> Self {
+        absurd
+    }
+
+    #[trusted]
+    #[logic]
+    fn addr_logic(self) -> Int {
+        absurd
+    }
+}
+
+pub trait GhostPtrExt<T: ?Sized>: Sized {
+    #[logic]
+    fn null_logic() -> Self;
+    #[logic]
+    fn addr_logic(self) -> Int;
+}
+
+extern_spec! {
+    impl<T> *const T {
+        // Safety since addr_logic is uninterpreted this just claims ptr::addr is deterministic
+        #[trusted]
+        #[ensures(result@ == self.addr_logic())]
+        fn addr(self) -> usize;
+
+        /// Check if `self` was created with [`GhostPtr::null`]
+        #[trusted]
+        #[ensures(result == (self == GhostPtr::<T>::null_logic()))]
+        fn is_null(self) -> bool;
+    }
+
+    mod std {
+        mod ptr {
+            /// Creates a null [`GhostPtr`] that no GhostToken has permission to
+            // Safety even though this pointer is dangling no GhostToken has permission to it so it's okay
+            #[trusted]
+            #[ensures(result == GhostPtr::<T>::null_logic())]
+            fn null<T>() -> *const T
+            where
+                T: Sized;
+        }
+    }
+}
+
+// Coverts `val` into a [`GhostPtr`] and gives `t` permission to it
+// Safety this pointer was owned by a box so no other GhostToken could have permission to it
+#[trusted]
+#[ensures(!(*t)@.contains(result))]
+#[ensures((^t)@ == (*t)@.insert(result, *val))]
+pub fn ptr_from_box<T>(val: Box<T>, t: &mut GhostPtrToken<T>) -> *const T {
+    Box::into_raw(val)
+}
+
+/// Immutably borrows `self`
+// Safety no other token has permission to `self`
+// `t` cannot be used to mutably borrow `self` as long as the shared lifetime lasts
+#[trusted]
+#[requires(t@.contains(ptr))]
+#[ensures(*result == t@.lookup(ptr))]
+pub fn ptr_as_ref<T>(ptr: *const T, t: &GhostPtrToken<T>) -> &T {
+    unsafe { &*ptr }
+}
+
+#[trusted]
+#[requires(new_model.subset((*t)@))]
+#[ensures((*result)@ == *new_model)]
+pub fn shrink_token_ref<T>(
+    t: &GhostPtrToken<T>,
+    new_model: Ghost<<GhostPtrToken<T> as ShallowModel>::ShallowModelTy>,
+) -> &GhostPtrToken<T> {
+    t
+}
+
+/// Mutably borrows `ptr` and shrinks `t` so that it can no longer be used to access `ptr`
+// Safety no other token has permission to `self`
+// `t` can no longer be used to access `ptr`
+#[trusted]
+#[requires((**t)@.contains(ptr))]
```

Comment:
I'm not a huge fan of having this mix of `**` `^*` `*^` and `^^`. I wonder if we can rephrase things to be more legible.

Reply:
I'm not sure how that would work, I need to refer to the 4 parts of the `&mut &mut GhostPtrToken`

Reply:
Yea, these kinds of code are just tricky, the comment makes the intent relatively clear